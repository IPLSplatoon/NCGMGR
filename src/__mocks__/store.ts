import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
