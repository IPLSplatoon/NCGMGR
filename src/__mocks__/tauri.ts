export const mockTauriEvent = {
    listen: jest.fn()
}

jest.mock('@tauri-apps/api/event', () => mockTauriEvent)

export const mockTauri = {
    invoke: jest.fn()
}

jest.mock('@tauri-apps/api/tauri', () => mockTauri)

export const mockTauriDialog = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/dialog', () => mockTauriDialog)

export const mockTauriFs = {
    readDir: jest.fn(),
    readTextFile: jest.fn(),
    removeDir: jest.fn(),
    removeFile: jest.fn(),
    createDir: jest.fn(),
    writeFile: jest.fn()
}

jest.mock('@tauri-apps/api/fs', () => mockTauriFs)

export const mockTauriOs = {
    type: jest.fn()
}

jest.mock('@tauri-apps/api/os', () => mockTauriOs)

export const mockTauriShell = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/shell', () => mockTauriShell)
