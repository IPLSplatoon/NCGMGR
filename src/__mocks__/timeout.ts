export function mockTimeout (): void {
    jest.spyOn(window, 'setTimeout').mockImplementation(handler => {
        // noinspection SuspiciousTypeOfGuard
        if (typeof handler !== 'string') {
            handler()
        }
        return 0
    })
}
