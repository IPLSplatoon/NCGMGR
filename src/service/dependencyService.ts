import { invoke } from '@tauri-apps/api/core'

export async function getNodejsVersion (): Promise<string | null> {
    return invoke('get_nodejs_version')
}
