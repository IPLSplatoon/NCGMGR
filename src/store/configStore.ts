import { defineStore } from 'pinia'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const useConfigStore = defineStore('config', {
    state: () => ({
        installPath: ''
    } as Configuration),
    actions: {
        load () {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            this.$patch(config)
        },
        persist () {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(this.$state))
        }
    }
})
