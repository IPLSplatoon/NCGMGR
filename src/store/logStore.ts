import { ActionState, LogEvent, ProgressEvent } from '@/types/log'
import { listen, UnlistenFn } from '@tauri-apps/api/event'
import { defineStore } from 'pinia'
import { listenForProcessExit } from '@/service/messagingService'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    progressEntries: Record<string, ProgressEvent>
    actionStates: Record<string, ActionState>
}

const unlistenFns: Record<string, UnlistenFn[]> = {}

export const useLogStore = defineStore('log', {
    state: () => ({
        lines: {},
        progressEntries: {},
        actionStates: {}
    } as LogStore),
    actions: {
        insertLine ({ line, key }: { line: LogEvent, key: string }) {
            if (!this.lines[key]) {
                this.lines[key] = []
            }
            this.lines[key].push(line)
        },
        setProgress (key: string, progress: ProgressEvent) {
            this.progressEntries[key] = progress
        },
        reset (key: string) {
            delete this.progressEntries[key]
            this.lines[key] = []
            this.actionStates[key] = ActionState.INCOMPLETE
        },
        setActionState ({ key, state }: { key: string, state: ActionState }) {
            this.actionStates[key] = state
        },
        listen (key: string, listenForProgress = false) {
            if (unlistenFns[key]) {
                return
            }

            const listenPromises: Array<Promise<UnlistenFn>> = []

            listenPromises.push(listen<LogEvent>(`log:${key}`, event => {
                this.insertLine({ line: event.payload, key })
            }))

            if (listenForProgress) {
                listenPromises.push(listen<ProgressEvent>(`progress:${key}`, event => {
                    this.setProgress(key, event.payload)
                }))
            }

            return Promise.all(listenPromises).then(result => {
                unlistenFns[key] = result
            })
        },
        unlisten (key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key].forEach(fn => fn())
                delete unlistenFns[key]
            }
        },
        logPromiseResult ({ promise, key }: { promise: Promise<unknown>, key: string }) {
            promise.catch(e => {
                this.insertLine({
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
                this.setActionState({ key, state: ActionState.COMPLETED_ERROR })
            })
        },
        async listenForProcessExit ({ key, callback }: { key: string, callback?: () => void }): Promise<void> {
            const unlisten = await listenForProcessExit(key, event => {
                this.setActionState({
                    key,
                    state: event.payload.success ? ActionState.COMPLETED_SUCCESS : ActionState.COMPLETED_ERROR
                })
                unlisten()
                if (callback) {
                    callback()
                }
            })
        }
    }
})
