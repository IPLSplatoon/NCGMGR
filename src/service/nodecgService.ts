import { mkdir, readDir, readTextFile, remove, writeFile } from '@tauri-apps/plugin-fs'
import { PackageSchema } from '@/types/package'
import isEmpty from 'lodash/isEmpty'
import { InstallStatus } from '@/store/nodecgStore'
import { invoke } from '@tauri-apps/api/core'
import { fileExists, folderExists } from '@/util/fs'
import { open } from '@tauri-apps/plugin-shell'
import { NodecgConfiguration } from '@/types/nodecg'
import { appLocalDataDir } from '@tauri-apps/api/path'

export async function getDefaultInstallDir (): Promise<string> {
    return `${await appLocalDataDir()}/nodecg`
}

export async function getNodecgStatus (directory: string | null): Promise<{ status: InstallStatus, message: string }> {
    if (directory == null || isEmpty(directory?.trim())) {
        return {
            status: InstallStatus.MISSING_INSTALL_DIRECTORY,
            message: 'Please select an installation directory.'
        }
    }

    const dir = await readDir(directory)
    if (dir.length < 1) {
        return {
            status: InstallStatus.READY_TO_INSTALL,
            message: 'Directory is empty. Ready to install...'
        }
    } else {
        const packageFile = dir.find(entry => entry.name === 'package.json')
        if (packageFile) {
            const packageJson: PackageSchema = JSON.parse(await readTextFile(`${directory}/package.json`))
            if (packageJson.name === 'nodecg') {
                return {
                    status: InstallStatus.INSTALLED,
                    message: `Found NodeCG v${packageJson.version}.`
                }
            } else {
                return {
                    status: InstallStatus.BAD_INSTALL_DIRECTORY,
                    message: `Found unknown package "${packageJson.name}".`
                }
            }
        } else {
            return {
                status: InstallStatus.BAD_INSTALL_DIRECTORY,
                message: 'Selected install directory is not empty, but NodeCG is not installed here.'
            }
        }
    }
}

export interface Bundle {
    name: string
    version?: string
}

export async function getBundles (directory: string | null): Promise<Bundle[]> {
    if (directory == null || isEmpty(directory.trim())) {
        throw new Error('No bundle directory provided.')
    }

    const bundlesDir = await readDir(directory + '/bundles')
    return Promise.all(
        bundlesDir
            .filter(entry => entry.isDirectory)
            .map(async (dir): Promise<Bundle | null> => {
                try {
                    const packageJson = JSON.parse(await readTextFile(`${directory}/bundles/${dir.name}/package.json`)) as PackageSchema
                    return {
                        name: packageJson.name,
                        version: packageJson.version == null || packageJson.version.trim() === '0.0.0'
                            ? await getBundleGitTag(packageJson.name)
                            : packageJson.version
                    }
                } catch (e) {
                    console.error('Failed to read package.json in bundle directory', e)
                    return null
                }
            }))
        .then((bundles) => bundles.filter(bundle => bundle != null) as Bundle[])
}

export async function getBundleVersions (bundleName: string): Promise<string[]> {
    return invoke('fetch_bundle_versions', { bundleName })
}

export async function configFileExists (bundleName: string, nodecgPath: string | null): Promise<boolean> {
    if (nodecgPath == null) {
        return false
    }
    return fileExists(`${nodecgPath}/cfg/${bundleName}.json`)
}

export async function removeBundle (bundleName: string, nodecgPath: string | null): Promise<[string, void]> {
    return Promise.all([
        invoke<string>('uninstall_bundle', { bundleName }),
        (async () => {
            if (await configFileExists(bundleName, nodecgPath)) {
                return remove(`${nodecgPath}/cfg/${bundleName}.json`)
            }
        })()
    ])
}

export async function openConfigFile (bundleName: string, nodecgPath: string | null): Promise<void> {
    if (nodecgPath == null) {
        return
    }

    return open(`${nodecgPath}/cfg/${bundleName}.json`)
}

export async function createConfigFile (bundleName: string, nodecgPath: string | null): Promise<void> {
    if (nodecgPath == null) {
        return
    }

    if (!await folderExists(`${nodecgPath}/cfg`)) {
        await mkdir(`${nodecgPath}/cfg`)
    }

    const encoder = new TextEncoder()
    return writeFile(`${nodecgPath}/cfg/${bundleName}.json`, encoder.encode('{\n\n}'))
}

export async function getNodecgConfig (nodecgPath: string): Promise<NodecgConfiguration | null> {
    if (!await folderExists(`${nodecgPath}/cfg`) || !await fileExists(`${nodecgPath}/cfg/nodecg.json`)) {
        return null
    }

    return JSON.parse(await readTextFile(`${nodecgPath}/cfg/nodecg.json`))
}

export async function openDashboard (nodecgPath: string): Promise<void> {
    const config = await getNodecgConfig(nodecgPath)
    return open(`http://localhost:${config?.port ?? '9090'}/dashboard`)
}

export async function getBundleGitTag (bundleName: string): Promise<string> {
    return invoke('get_bundle_git_tag', { bundleName })
}
