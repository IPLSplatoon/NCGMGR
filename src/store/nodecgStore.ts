import { Bundle, getBundles, getNodecgStatus } from '@/service/nodecgService'
import { useConfigStore } from '@/store/configStore'
import { defineStore } from 'pinia'
import { listenForProcessExit } from '@/service/messagingService'

export enum InstallStatus {
    UNKNOWN,
    READY_TO_INSTALL,
    INSTALLED,
    UNABLE_TO_INSTALL,
    MISSING_INSTALL_DIRECTORY,
    BAD_INSTALL_DIRECTORY
}

export enum RunStatus {
    NOT_STARTED = 'NOT_STARTED',
    RUNNING = 'RUNNING',
    STOPPED = 'STOPPED'
}

export interface NodecgStore {
    status: {
        installStatus: InstallStatus
        runStatus: RunStatus
        message: string
        bundlesLoading: boolean
    },
    bundles: Bundle[]
}

export const useNodecgStore = defineStore('nodecg', {
    state: () => ({
        status: {
            installStatus: InstallStatus.UNKNOWN,
            runStatus: RunStatus.NOT_STARTED,
            message: '',
            bundlesLoading: false
        },
        bundles: []
    } as NodecgStore),
    actions: {
        async checkNodecgStatus () {
            const configStore = useConfigStore()

            try {
                const { status, message } = await getNodecgStatus(configStore.userConfig.nodecgInstallDir)
                this.status.message = message
                this.status.installStatus = status
                if (status === InstallStatus.INSTALLED) {
                    this.getBundleList()
                } else {
                    this.bundles = []
                }
            } catch (e) {
                this.status.message = String(e)
                this.status.installStatus = InstallStatus.UNABLE_TO_INSTALL
                this.bundles = []
            }
        },
        async getBundleList () {
            const configStore = useConfigStore()
            this.status.bundlesLoading = true
            try {
                this.bundles = await getBundles(configStore.userConfig.nodecgInstallDir)
            } finally {
                this.status.bundlesLoading = false
            }
        }
    }
})

listenForProcessExit('run-nodecg', () => {
    useNodecgStore().$state.status.runStatus = RunStatus.STOPPED
})
