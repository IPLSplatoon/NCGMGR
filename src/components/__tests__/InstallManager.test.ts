import { mockTauri, mockTauriDialog } from '@/__mocks__/tauri'
import { config, flushPromises, shallowMount } from '@vue/test-utils'
import InstallManager from '../InstallManager.vue'
import { InstallStatus, useNodecgStore, RunStatus } from '@/store/nodecgStore'
import { IplButton } from '@iplsplatoon/vue-components'
import { createTestingPinia, TestingPinia } from '@pinia/testing'
import { useConfigStore } from '@/store/configStore'
import { useLogStore } from '@/store/logStore'
import { openDashboard } from '@/service/nodecgService'

jest.mock('@/service/nodecgService')

describe('Installer', () => {
    config.global.stubs = {
        IplSpace: false,
        IplExpandingSpace: false
    }

    let pinia: TestingPinia

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
    })

    it('matches snapshot', () => {
        useConfigStore().installPath = '/install/path'

        const wrapper = shallowMount(InstallManager)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it.each(Object.values(RunStatus))('matches snapshot when nodecg status is %s', (status: RunStatus) => {
        useNodecgStore().status.runStatus = status

        expect(shallowMount(InstallManager).html()).toMatchSnapshot()
    })

    it('handles directory selection', async () => {
        const store = useConfigStore()
        store.persist = jest.fn()
        const wrapper = shallowMount(InstallManager)
        mockTauriDialog.open.mockResolvedValue('/new/path')

        wrapper.getComponent<typeof IplButton>('[data-test="install-directory-select-button"]').vm.$emit('click')
        await flushPromises()

        expect(mockTauriDialog.open).toHaveBeenCalledWith({ directory: true })
        expect(store.installPath).toEqual('/new/path')
        expect(store.persist).toHaveBeenCalled()
    })

    it('handles directory selection getting cancelled', async () => {
        const store = useConfigStore()
        store.persist = jest.fn()
        const wrapper = shallowMount(InstallManager)
        mockTauriDialog.open.mockResolvedValue(null)

        wrapper.getComponent<typeof IplButton>('[data-test="install-directory-select-button"]').vm.$emit('click')
        await flushPromises()

        expect(store.installPath).toEqual('')
        expect(store.persist).not.toHaveBeenCalled()
    })

    it('disables installation if installation is not possible', async () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.checkNodecgStatus = jest.fn()
        nodecgStore.status.installStatus = InstallStatus.UNABLE_TO_INSTALL
        const wrapper = await shallowMount(InstallManager)

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('true')
        expect(wrapper.findComponent('[data-test="launch-button"]').exists()).toEqual(false)
    })

    it('enables installation if installation is possible', async () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.checkNodecgStatus = jest.fn()
        nodecgStore.status.installStatus = InstallStatus.READY_TO_INSTALL
        const wrapper = await shallowMount(InstallManager)

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('false')
        expect(wrapper.findComponent('[data-test="launch-button"]').exists()).toEqual(false)
    })

    it('shows start/stop button if nodecg is installed', async () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.checkNodecgStatus = jest.fn()
        nodecgStore.status.installStatus = InstallStatus.INSTALLED
        const wrapper = await shallowMount(InstallManager)

        expect(wrapper.findComponent('[data-test="install-button"]').exists()).toEqual(false)
        expect(wrapper.findComponent('[data-test="start-stop-toggle-button"]').exists()).toEqual(true)
    })

    it('disables installation if installation status is unknown', async () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.status.installStatus = InstallStatus.UNKNOWN
        const wrapper = await shallowMount(InstallManager)
        await flushPromises()

        expect(wrapper.getComponent('[data-test="install-button"]').attributes().disabled).toEqual('true')
        expect(wrapper.findComponent('[data-test="launch-button"]').exists()).toEqual(false)
    })

    it('handles installation', async () => {
        const configStore = useConfigStore()
        configStore.installPath = '/install/path'
        const logStore = useLogStore()
        logStore.reset = jest.fn()
        logStore.listen = jest.fn()
        logStore.logPromiseResult = jest.fn()
        logStore.listenForProcessExit = jest.fn()
        const nodecgStore = useNodecgStore()
        nodecgStore.status.installStatus = InstallStatus.READY_TO_INSTALL
        mockTauri.invoke.mockResolvedValue({})
        const wrapper = shallowMount(InstallManager)

        await wrapper.getComponent<typeof IplButton>('[data-test="install-button"]').vm.$emit('click')

        expect(logStore.reset).toHaveBeenCalledWith('install-nodecg')
        expect(logStore.listen).toHaveBeenCalledWith('install-nodecg')
        expect(mockTauri.invoke).toHaveBeenCalledWith('install_nodecg', { path: '/install/path' })
        expect(logStore.logPromiseResult).toHaveBeenCalledWith({ promise: expect.anything(), key: 'install-nodecg' })
        expect(logStore.listenForProcessExit).toHaveBeenCalledWith({ key: 'install-nodecg', callback: expect.anything() })
    })

    it.each([RunStatus.NOT_STARTED, RunStatus.STOPPED])('handles launching when nodecg status is %s', async (status) => {
        const configStore = useConfigStore()
        configStore.installPath = '/install/path'
        const logStore = useLogStore()
        logStore.reset = jest.fn()
        logStore.logPromiseResult = jest.fn()
        const nodecgStore = useNodecgStore()
        nodecgStore.status.installStatus = InstallStatus.INSTALLED
        nodecgStore.status.runStatus = status
        mockTauri.invoke.mockResolvedValue({})
        const wrapper = shallowMount(InstallManager)

        wrapper.getComponent<typeof IplButton>('[data-test="start-stop-toggle-button"]').vm.$emit('click')
        await flushPromises()

        expect(logStore.reset).toHaveBeenCalledWith('run-nodecg')
        expect(mockTauri.invoke).toHaveBeenCalledWith('start_nodecg', { path: '/install/path' })
        expect(logStore.logPromiseResult).toHaveBeenCalledWith({ promise: expect.anything(), key: 'run-nodecg' })
        expect(nodecgStore.status.runStatus).toEqual(RunStatus.RUNNING)
    })

    it('handles stopping nodecg', async () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.status.installStatus = InstallStatus.INSTALLED
        nodecgStore.status.runStatus = RunStatus.RUNNING
        mockTauri.invoke.mockResolvedValue({})
        const wrapper = shallowMount(InstallManager)

        await wrapper.getComponent<typeof IplButton>('[data-test="start-stop-toggle-button"]').vm.$emit('click')

        expect(mockTauri.invoke).toHaveBeenCalledWith('stop_nodecg')
    })

    it('handles opening the dashboard', () => {
        useConfigStore().installPath = '/nodecg/install/path'
        const wrapper = shallowMount(InstallManager)

        wrapper.getComponent<typeof IplButton>('[data-test="open-dashboard-button"]').vm.$emit('click')

        expect(openDashboard).toHaveBeenCalledWith('/nodecg/install/path')
    })
})
