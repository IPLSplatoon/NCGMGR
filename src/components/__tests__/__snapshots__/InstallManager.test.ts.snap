// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Installer matches snapshot 1`] = `
<div class="ipl-space color-dark layout vertical" data-v-11cc5f1c="">
  <div>Installation folder: /install/path</div>
  <status-row-stub color="gray" class="m-t-8"></status-row-stub>
  <div class="layout horizontal m-t-8">
    <ipl-button-stub label="Select folder" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-directory-select-button"></ipl-button-stub>
    <ipl-button-stub label="Install" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-button" class="m-l-8"></ipl-button-stub>
    <ipl-button-stub label="Open dashboard" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="open-dashboard-button" class="m-l-8"></ipl-button-stub>
  </div>
</div>
<div class="ipl-expansion-panel__content m-t-8" data-test="nodecg-log-space" data-v-67856a23="" style="display: none;">
  <div class="ipl-expansion-panel__header" data-v-67856a23="">
    <div class="ipl-expansion-panel__header-background" data-v-67856a23=""></div>
    <div class="ipl-expansion-panel__title" data-v-67856a23=""> NodeCG log <span class="badge badge-red">Stopped</span></div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon content-expanded" data-v-67856a23=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-67856a23=""></div>
  </div>
  <div class="content" data-v-67856a23="">
    <div class="layout vertical center-horizontal">
      <log-display-stub logkey="run-nodecg" class="m-t-4"></log-display-stub>
    </div>
  </div>
</div>
<log-overlay-stub title="Installing..." logkey="install-nodecg" visible="false" nobackgroundclose="true" withprogress="true"></log-overlay-stub>
`;

exports[`Installer matches snapshot when nodecg status is NOT_STARTED 1`] = `
<div class="ipl-space color-dark layout vertical" data-v-11cc5f1c="">
  <div>Installation folder: </div>
  <status-row-stub color="gray" class="m-t-8"></status-row-stub>
  <div class="layout horizontal m-t-8">
    <ipl-button-stub label="Select folder" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-directory-select-button"></ipl-button-stub>
    <ipl-button-stub label="Install" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-button" class="m-l-8"></ipl-button-stub>
    <ipl-button-stub label="Open dashboard" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="open-dashboard-button" class="m-l-8"></ipl-button-stub>
  </div>
</div>
<div class="ipl-expansion-panel__content m-t-8" data-test="nodecg-log-space" data-v-67856a23="" style="display: none;">
  <div class="ipl-expansion-panel__header" data-v-67856a23="">
    <div class="ipl-expansion-panel__header-background" data-v-67856a23=""></div>
    <div class="ipl-expansion-panel__title" data-v-67856a23=""> NodeCG log <span class="badge badge-red">Stopped</span></div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon content-expanded" data-v-67856a23=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-67856a23=""></div>
  </div>
  <div class="content" data-v-67856a23="">
    <div class="layout vertical center-horizontal">
      <log-display-stub logkey="run-nodecg" class="m-t-4"></log-display-stub>
    </div>
  </div>
</div>
<log-overlay-stub title="Installing..." logkey="install-nodecg" visible="false" nobackgroundclose="true" withprogress="true"></log-overlay-stub>
`;

exports[`Installer matches snapshot when nodecg status is RUNNING 1`] = `
<div class="ipl-space color-dark layout vertical" data-v-11cc5f1c="">
  <div>Installation folder: </div>
  <status-row-stub color="gray" class="m-t-8"></status-row-stub>
  <div class="layout horizontal m-t-8">
    <ipl-button-stub label="Select folder" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-directory-select-button"></ipl-button-stub>
    <ipl-button-stub label="Install" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-button" class="m-l-8"></ipl-button-stub>
    <ipl-button-stub label="Open dashboard" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="open-dashboard-button" class="m-l-8"></ipl-button-stub>
  </div>
</div>
<div class="ipl-expansion-panel__content m-t-8" data-test="nodecg-log-space" data-v-67856a23="">
  <div class="ipl-expansion-panel__header" data-v-67856a23="">
    <div class="ipl-expansion-panel__header-background" data-v-67856a23=""></div>
    <div class="ipl-expansion-panel__title" data-v-67856a23=""> NodeCG log <span class="badge badge-green">Running</span></div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon content-expanded" data-v-67856a23=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-67856a23=""></div>
  </div>
  <div class="content" data-v-67856a23="">
    <div class="layout vertical center-horizontal">
      <log-display-stub logkey="run-nodecg" class="m-t-4"></log-display-stub>
    </div>
  </div>
</div>
<log-overlay-stub title="Installing..." logkey="install-nodecg" visible="false" nobackgroundclose="true" withprogress="true"></log-overlay-stub>
`;

exports[`Installer matches snapshot when nodecg status is STOPPED 1`] = `
<div class="ipl-space color-dark layout vertical" data-v-11cc5f1c="">
  <div>Installation folder: </div>
  <status-row-stub color="gray" class="m-t-8"></status-row-stub>
  <div class="layout horizontal m-t-8">
    <ipl-button-stub label="Select folder" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-directory-select-button"></ipl-button-stub>
    <ipl-button-stub label="Install" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="install-button" class="m-l-8"></ipl-button-stub>
    <ipl-button-stub label="Open dashboard" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="open-dashboard-button" class="m-l-8"></ipl-button-stub>
  </div>
</div>
<div class="ipl-expansion-panel__content m-t-8" data-test="nodecg-log-space" data-v-67856a23="">
  <div class="ipl-expansion-panel__header" data-v-67856a23="">
    <div class="ipl-expansion-panel__header-background" data-v-67856a23=""></div>
    <div class="ipl-expansion-panel__title" data-v-67856a23=""> NodeCG log <span class="badge badge-red">Stopped</span></div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon content-expanded" data-v-67856a23=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-67856a23=""></div>
  </div>
  <div class="content" data-v-67856a23="">
    <div class="layout vertical center-horizontal">
      <log-display-stub logkey="run-nodecg" class="m-t-4"></log-display-stub>
    </div>
  </div>
</div>
<log-overlay-stub title="Installing..." logkey="install-nodecg" visible="false" nobackgroundclose="true" withprogress="true"></log-overlay-stub>
`;
